[NFC][ADT] Remove << operators from StringRefTest

jurahul opened the pull request:
- Remove ostream << operators for StringRef and StringRef pair from StringTest. 
  Both of these are natively supported by googletest framework.

s-barannikov commented on llvm/unittests/ADT/StringRefTest.cpp:
```diff
@@ -16,20 +16,7 @@
 
-namespace llvm {
-
-std::ostream &operator<<(std::ostream &OS, const StringRef &S) {
```
These are not unused.
They are used by gtest on a test failure to print human readable representation of StringRefs.
They must be in llvm namespace.

They *should* be moved to a place where *all* GTests can see them, otherwise there is an ODR violation as explained [here](https://stackoverflow.com/questions/24673515/googletest-printto-not-getting-called-for-a-class) and they may fail to work.

jurahul replied:
Thanks. I guess it this part:

>This has caught me out, so I'm really glad to find this solution. My case was that MyType has an ostream& operator<<(), but it's defined in mytype_io.h rather than mytype.h. If any unit test cpp file uses MyType but doesn't include mytype_io.h then it will cause the default template implementation in gtest-printers.h to be instantiated.

>So my type wasn't being printed correctly. The fix was to ensure that all unit tests that used MyType included mytype_io.h.

So without the << operators, it will call some default printers which won't print the StringRef the right way. May be I can add a comment to that effect?

s-barannikov replied:
> So without the << operators, it will call some default printers which won't print the StringRef the right way.

They may or may not work, depending on which instantiation is picked by the linker.

> May be I can add a comment to that effect?

A comment would be great, thanks.

jurahul replied:
Hmm, I can see that even without these, the StringRef and the pair gets printed properly:

```
[ RUN      ] StringRefTest.Construction
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/StringRefTest.cpp:41: Failure
Expected equality of these values:
  ""
    Which is: 0x55db55e939d4
  StringRef("a")
    Which is: "a"
```

and

```
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/StringRefTest.cpp:166: Failure
Expected equality of these values:
  std::make_pair(StringRef("he"), StringRef("lo"))
    Which is: ("he", "lo")
  Str.split('y')
    Which is: ("hello", "")
```

jurahul replied:
SG, I'll restore. In any case its an overreach for what the change claims to do.

s-barannikov replied:
Thanks for checking. This is probably because there is another printer [here](https://github.com/llvm/llvm-project/blob/a866ce789eb99da4d7a486eeb60a53be6c75f4fd/third-party/unittest/googletest/include/gtest/internal/custom/gtest-printers.h#L52) and it takes precedence.

I wasn't aware of this header until now. I think the overloaded << for StringRef can safely be removed then.
(and I guess `pair<StringRef, StringRef>` overload is still needed)

jurahul replied:
Doh! I'll remove them again :) What about pair<>? Is there a generic pair printer in the test framework?

jurahul replied:
may be its [this ](https://github.com/google/googletest/blob/main/googletest/include/gtest/gtest-printers.h#L812)for pair.

jurahul replied:
Restored deleting of << operators.

jurahul replied:
Updated change description as well to say:

* Delete unused << operators (googltest framework provides printers for StringRef and pairs).

s-barannikov commented on llvm/unittests/ADT/StringRefTest.cpp:
```diff
@@ -47,37 +34,39 @@ static_assert(std::is_assignable_v<StringRef &, const char *&&>,
-namespace {
-TEST(StringRefTest, Construction) {
+// Test fixture.
+class StringRefTest : public testing::Test {};
```
Why do you need it with no methods?

jurahul replied:
Because that will come when I add test cases for reverse iterators, in a follow-on change. I assume I should not be rolling in this change into that. And I see other tests, like StringSetTest using a harness without any members.

s-barannikov reviewed (approved):
LGTM but please give others a chance to review. I have little experience with gtest.

jurahul commented:
> give others a chance to review

Of course. Will wait for @kuhar

jurahul commented:
Thanks for the review!

kuhar reviewed (commented):
@jurahul could you explain why reverse iterator tests will need this test class?

Even if they do, we could make them use their test harness and leave the existing tests use test functions.

jurahul commented:
> @jurahul could you explain why reverse iterator tests will need this test class?
> 
> Even if they do, we could make them use their test harness and leave the existing tests use test functions.

I want to use a helper template function in the test harness to help test all 4 flavors of iterators:

```
 template <bool Const, bool Forward>
 void TestIteration(void) {
   using StringRefTy = std::conditional_t<Const, const StringRef, StringRef>;
   StringRefTy S("Hello");

   auto [Begin, End] = [S]() {
     if constexpr (Forward)
       return std::make_pair(S.begin(), S.end());
     else
       return std::make_pair(S.rbegin(), S.rend());
   }();
   const char *Reference = Forward ? "Hello" : "olleH";
   for (auto II = Begin; II != End; ++II, ++Reference)
     EXPECT_EQ(*II, *Reference);
 };
```

I could add a harness just for this subtest, but then I am assuming it will be something like StringRefIteratorTest. and then the test name will be "StringRefIteratorTest.Forward" etc. I was attempting to keep the names uniform, but in hindsight that may not be important. OTOH, I do see other tests that use test harness, and when they do, it seems to be used uniformly throughout the test, irrespective of whether the subtest actually uses anything from the harness. For example, StringSetTest has an empty harness, and SmallStringTest has a harness and some tests that do not use anything from the harness.

I guess having a single harness helps keep the test name same throughout the test, but that's not a requirement. I am assuming if there is atest harness, you cannot use TEST(SameNameAsHarness).

kuhar commented:
To me this seems like overkill for what we need to check that rbegin/rend point to the right place... For example, you can use `llvm::equal` to check that ranges match.

jurahul commented:
> To me this seems like overkill for what we need to check that rbegin/rend point to the right place... For example, you can use `llvm::equal` to check that ranges match.

SG. I'll repurpose this to just remove the << operators which we have discovered as not needed.

jurahul committed:
[NFC][ADT] Remove << operators from StringTest

- Remove ostream << operators for StringRef and StringRef pair
  from StringTest. Both of these are natively supported by
  googletest framework

jurahul commented:
@kuhar PTAL again.
